### PR TITLE
M0078DDG-344 Document NLS External Loader

### PR DESCRIPTION
**Pull request by gjadi-mej**
https://docs.microej.com/en/feature-m0078ddg-344_nls-external-loader/ApplicationDeveloperGuide/nlsExternalLoader.html#nls-external-loader

### CODE REVIEW

File: ApplicationDeveloperGuide/nlsExternalLoader.rst [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+NLS External Loader
+===================
+
+MicroEJ provides Native Language Support in applications as described in the :ref:`section.applicationResources.nls` section.
+Languages strings are defined in PO files in the application sources.
+While this is convenient during the application development phase since everything is at the same location,
+it can be painful in others circumstances.
+For example, during the test of the application or the translation phase by non-developer persons, 
+it would mean asking the developers to update the sources filesand rebuild the application for each update.
```
Comment (gjadi-mej) [inline ApplicationDeveloperGuide/nlsExternalLoader.rst:11]:
```suggestion
```

File: ApplicationDeveloperGuide/nlsExternalLoader.rst [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+In order to make NLS changes easier and quicker, PO files can be updated directly in a Virtual Device thanks to the NLS External Loader.
+PO files can be dropped in a given location in the Virtual Device folders to dynamically replace the language strings embedded in the application.
+
+Installation
+------------
+
+To enable the NLS External Loader in your Virtual Device, add the following dependency to the `module.ivy` file of your MicroEJ Firmware project:
```
Comment (gjadi-mej) [inline ApplicationDeveloperGuide/nlsExternalLoader.rst:19]:
```suggestion
To enable the NLS External Loader in the Virtual Device, add the following dependency to the ``module.ivy`` file of the MicroEJ Firmware project:
```

File: ApplicationDeveloperGuide/nlsExternalLoader.rst [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+.. code-block:: xml
+
+   <dependency org="com.microej.tool" name="nls-po-external-loader" rev="2.3.0" transitive="false"/>
+
+Then rebuild your MicroEJ Firmware project to produce the Virtual Device.
```
Comment (gjadi-mej) [inline ApplicationDeveloperGuide/nlsExternalLoader.rst:25]:
```suggestion
Then rebuild the MicroEJ Firmware project to produce the Virtual Device.
```

File: ApplicationDeveloperGuide/nlsExternalLoader.rst [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+Usage
+-----
+
+Once the project built:
+
+- unzip the Virtual Device and create a folder named ``translations`` in the root folder.
+- copy all the PO files from your project into the ``translations`` folder.
```
Comment (gjadi-mej) [inline ApplicationDeveloperGuide/nlsExternalLoader.rst:33]:
```suggestion
- copy all the PO files from the project into the ``translations`` folder.
```

File: ApplicationDeveloperGuide/nlsExternalLoader.rst [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  All PO files found in this folder are processed, no matter their folder level.
+- start the Virtual Device with the launcher.
+  You can check in the logs that the NLS External Loader has been executed and has found the PO files::
```
Comment (gjadi-mej) [inline ApplicationDeveloperGuide/nlsExternalLoader.rst:36]:
```suggestion
- start the Virtual Device with the launcher.
   The following logs should be printed if the NLS External Loader has been executed and has found the PO files::
```

File: ApplicationDeveloperGuide/nlsExternalLoader.rst [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+      [mkdir] Created dir: <PATH>\tmp\microejlaunch1307817858\resourcebuffer
+   [po-to-nls] *.nls files found in <PATH>\output\<FIRMWARE>\resourceBuffer :
+   [po-to-nls]   - <CLASSPATH>.<CLASSNAME>
+   [po-to-nls]   - <CLASSPATH>.<CLASSNAME>
+   [po-to-nls] Loading *.po files for NLS interface f<CLASSPATH>.<CLASSNAME>
+   [po-to-nls]   => loaded locales : fr_DA,fr_ES,fr_NL,fr_FI,fr_RU,fr_SV,fr_DE,fr_NO,fr_IT,fr_FR,fr_PL,fr_EN
+   [po-to-nls] Loading *.po files for NLS interface <CLASSPATH>.<CLASSNAME>
+   [po-to-nls]   => loaded locales : fr_DA,fr_ES,fr_NL,fr_FI,fr_RU,fr_SV,fr_DE,fr_NO,fr_IT,fr_FR,fr_PL,fr_EN
+
+- update the languages strings in your PO files.
```
Comment (gjadi-mej) [inline ApplicationDeveloperGuide/nlsExternalLoader.rst:50]:
```suggestion
- update the languages strings in the PO files of the Virtual Device (the files in the `translations/` folder).
```

File: ApplicationDeveloperGuide/nlsExternalLoader.rst [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+- restart the Virtual Device and check your changes.
```
Comment (gjadi-mej) [inline ApplicationDeveloperGuide/nlsExternalLoader.rst:51]:
```suggestion
- restart the Virtual Device and check the changes.
```

File: ApplicationDeveloperGuide/nlsExternalLoader.rst [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+It is important to know the following rules about the NLS External Loader:
+
+- the external PO files names must match with the default PO files names of the application to be processed.
+- when PO files with a given name are loaded, the default translations for these PO files are replaced, there is no merge. It means that:
+
+  - if messages are missing in the new PO files, they are not available anymore for the application and may very probably make it crash.
+  - if languages are missing (the application has 3 PO files for English, French and Spanish, 
+    and only PO files for English and French are available in the translations folder), 
+    the messages of the missing languages arenot available anymore for the application and may very probably make it crash.
```
Comment (gjadi-mej) [inline ApplicationDeveloperGuide/nlsExternalLoader.rst:62]:
```suggestion
    the messages of the missing languages are not available anymore for the application and may very probably make it crash.
```

File: ApplicationDeveloperGuide/nlsExternalLoader.rst [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+  - if new messages are added in the PO files, it has no impact, they are just not used in the application.
```
Comment (gjadi-mej) [inline ApplicationDeveloperGuide/nlsExternalLoader.rst:63]:
```suggestion
  - if new messages are added in the PO files, it has no impact, they are ignored by the application.
```

File: ApplicationDeveloperGuide/nlsExternalLoader.rst [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+
+Crowdin
+-------
+
+Crowdin is a cloud-based localization platform which allows to manage multilingual content.
+The NLS External Loader can fetch translations directly from Crowdin to make the translation process even easier.
+Translators can then contribute and validate their translations in Crowdin and apply them automatically in the Virtual Device.
+
+A new dependency must be added to the `module.ivy` file of your MicroEJ Firmware project to enable this integration:
```
Comment (gjadi-mej) [inline ApplicationDeveloperGuide/nlsExternalLoader.rst:96]:
```suggestion
A new dependency must be added to the `module.ivy` file of the MicroEJ Firmware project to enable this integration:
```

File: ApplicationDeveloperGuide/nlsExternalLoader.rst [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+
+.. code-block:: xml
+
+   <dependency org="com.microej.tool" name="nls-po-crowdin" rev="1.0.0" transitive="false"/>
+
+Once the module has been built, edit the file ``platform/tools/crowdin/crowdin.properties`` to configure the Crowdin connection:
+
+- set ``crowdin.token`` to your Crowdin API token. A token can be generated in the Crowdin in ``Profile > Settings > ??? > Add token``.
```
Comment (gjadi-mej) [inline ApplicationDeveloperGuide/nlsExternalLoader.rst:104]:
```suggestion
- set ``crowdin.token`` to the Crowdin API token. A token can be generated in the Crowdin in :guilabel:`Settings` > :guilabel:`API` > click on :guilabel:`New Token`.
```

File: ApplicationDeveloperGuide/nlsExternalLoader.rst [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+- set ``crowdin.projectsIds`` to the id of your Crowdin project. The project id can be found in ``??? > ??? > ???``.
```
Comment (gjadi-mej) [inline ApplicationDeveloperGuide/nlsExternalLoader.rst:105]:
```suggestion
- set ``crowdin.projectsIds`` to the id of a Crowdin project. The project id can be found in the :guilabel:`Details` section on a project page.
```

File: ApplicationDeveloperGuide/nlsExternalLoader.rst [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+  Multiple projects can set by separating their id with a comma (for example ``crowdin.projectsIds=12,586,874``).
```
Comment (gjadi-mej) [inline ApplicationDeveloperGuide/nlsExternalLoader.rst:106]:
```suggestion
  Multiple projects can be set by separating their id with a comma (for example ``crowdin.projectsIds=12,586,874``).
```

File: ApplicationDeveloperGuide/nlsExternalLoader.rst [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+
+When the configuration is done, the fetch of the Crowdin translations can be done by executing the script ``crowdin.bat`` or ``crowdin.sh``.
```
Comment (gjadi-mej) [inline ApplicationDeveloperGuide/nlsExternalLoader.rst:108]:
```suggestion
When the configuration is done, the fetch of the Crowdin translations can be done by executing the script ``crowdin.bat`` or ``crowdin.sh`` located in the folder ``platform/tools/crowdin/``.
```

File: ApplicationDeveloperGuide/nlsExternalLoader.rst (hunk 3 above)
Comment (tdelhomenie-microej) [reply ApplicationDeveloperGuide/nlsExternalLoader.rst:25]: Is there an official rule about the usage of "you" ? It is used in multiple places in the doc, for example in https://docs.microej.com/en/latest/ApplicationDeveloperGuide/UI/MicroUI/fontDesigner.html.
I can only find this ticket on this topic : https://youtrack.cross/issue/M0078DDG-288 but not sure of its status.

File: ApplicationDeveloperGuide/nlsExternalLoader.rst (hunk 1 above)
Comment (tdelhomenie-microej) [reply ApplicationDeveloperGuide/nlsExternalLoader.rst:11]: You want to delete these lines ?

File: ApplicationDeveloperGuide/nlsExternalLoader.rst (hunk 5 above)
Comment (tdelhomenie-microej) [reply ApplicationDeveloperGuide/nlsExternalLoader.rst]: Done.

File: ApplicationDeveloperGuide/nlsExternalLoader.rst (hunk 8 above)
Comment (tdelhomenie-microej) [reply ApplicationDeveloperGuide/nlsExternalLoader.rst]: Done.

File: ApplicationDeveloperGuide/nlsExternalLoader.rst (hunk 9 above)
Comment (tdelhomenie-microej) [reply ApplicationDeveloperGuide/nlsExternalLoader.rst]: Done.

File: ApplicationDeveloperGuide/nlsExternalLoader.rst (hunk 13 above)
Comment (tdelhomenie-microej) [reply ApplicationDeveloperGuide/nlsExternalLoader.rst]: Done.

File: ApplicationDeveloperGuide/nlsExternalLoader.rst (hunk 14 above)
Comment (tdelhomenie-microej) [reply ApplicationDeveloperGuide/nlsExternalLoader.rst]: Done.

**Pull request merged**